### PR TITLE
Fix some parsing issues, works with premium chapters

### PR DESCRIPTION
**Pull request by TheUnlocked**

### CODE REVIEW

**Comment by alessandrojean** [comment]
Have you logged in with a premium account through WebView and tested if it works?

**Comment by TheUnlocked** [comment]
I have. I've also tried logging out and it successfully doesn't work in that case.

**Comment by alessandrojean** [comment]
> I have. I've also tried logging out and it successfully doesn't work in that case.

Probably because the response was cached, maybe trying to force network on the OkHttp call could help, like this:

https://github.com/inorichi/tachiyomi-extensions/blob/0dc21c1b38113a647beb143a4e53818425d4c5bf/src/all/mangadex/src/eu/kanade/tachiyomi/extension/all/mangadex/MangaDex.kt#L115

File: src/en/vizshonenjump/src/eu/kanade/tachiyomi/extension/en/vizshonenjump/VizShonenJump.kt [hunk 1]
```diff
@@ -134,7 +134,14 @@ class VizShonenJump : ParsedHttpSource() {
         }
 
         scanlator = "VIZ Media"
-        url = element.attr("href")
+
+        url = element.attr("data-target-url")
+        if (url.startsWith("javascript")) {
```
Comment (alessandrojean) [inline src/en/vizshonenjump/src/eu/kanade/tachiyomi/extension/en/vizshonenjump/VizShonenJump.kt:139]: Couldn't this block be like this?

```kotlin
if (url.startsWith("javascript:tryReadChapter") {
  url = url.substringAfterFirst("'").substringAfterLast("'")
} else {
  name += " (Free)"
}
```

**Comment by TheUnlocked** [comment]
> > I have. I've also tried logging out and it successfully doesn't work in that case.
> 
> Probably because the response was cached, maybe trying to force network on the OkHttp call could help, like this:
> 
> https://github.com/inorichi/tachiyomi-extensions/blob/0dc21c1b38113a647beb143a4e53818425d4c5bf/src/all/mangadex/src/eu/kanade/tachiyomi/extension/all/mangadex/MangaDex.kt#L115

Still works properly.

The simplification of the premium chapter parsing seems to work fine.

**Comment by alessandrojean** [comment]
There's just one issue left: the ideal should be only the free/available without login chapters appearing to the non logged in users.

I think this can be done with a modification to the `chapterListParse` and `chapterFromElement` methods:

```kotlin
override fun chapterListParse(response: Response): List<SChapter> {
    return super.chapterListParse(response)
        .filter { it.url != "LOCKED" } 
}

override fun chapterFromElement(element: Element): SChapter = SChapter.create().apply {
    // ...
    url = element.attr("data-target-url")
    if (url.startsWith("javascript:tryReadChapter")) {
        url = url.substringAfter("'").substringBeforeLast("'")
    } else if (url.startsWith("javascript") {
        url = "LOCKED"
    }
}
```

I also think that maybe there's no need to append the `(Free)` to the chapter name if this works.

Can you check if this approach works as expected? 

*To show only the free chapters if you are not logged in, and show all the chapters (after a refresh) if you are logged in.*

**Comment by TheUnlocked** [comment]
That's what I was originally going for, but unfortunately the site loads with every button locked and unlocks them via javascript. It might be possible to detect if the user has access through cookie inspection though... I'll look into that.

**Comment by alessandrojean** [comment]
Oh, that's bad. So, let as it was, but do these modifications:

```kotlin
// ...
    url = element.attr("data-target-url")
    if (url.startsWith("javascript:tryReadChapter")) {
        url = url.substringAfter("'").substringBeforeLast("'")
    } else if (url.startsWith("javascript") {
        name = "[🔒] " + name
        url = "LOCKED"
    } else {
       name = "[FREE] " + name
   }
}

override fun fetchPageList(chapter: SChapter): Observable<List<Page> {
    if (chapter.url == "LOCKED") {
        return Observable.error(Exception("You need to login to read this chapter"))
    }

    return super.fetchPageList(chapter)
}
```

Then it should be good.

**Comment by alessandrojean** [comment]
> That's what I was originally going for, but unfortunately the site loads with every button locked and unlocks them via javascript. It might be possible to detect if the user has access through cookie inspection though... I'll look into that.

Urrgh, but since it unlocks via javascript, the code above probably will not work properly either. 

When you are not logged in, what is the URL of the locked chapters? When you log in, they change to `javascript:tryReadChapter` or they are like this even if you are not logged in?

**Comment by TheUnlocked** [comment]
> When you are not logged in, what is the URL of the locked chapters? When you log in, they change to `javascript:tryReadChapter` or they are like this even if you are not logged in?

The URLs are the same as when you're not logged in, but they change via javascript to the content of the `data-target-url` attribute. If we're comfortable making the assumption that anyone who's logged in will have a subscription, we may be able to do it that way (though I'm not sure-- I'm experimenting with it right now).

**Comment by alessandrojean** [comment]
Is the `data-target-url` attribute present when the user is not logged in?

**Comment by TheUnlocked** [comment]
Yes.

**Comment by alessandrojean** [comment]
And the value is the same to logged in and not logged in?

What happens when you try to access a locked chapter directly by the URL when you are not logged in? It returns some HTTP Error Code, redirects you to another page? If so, at least the `"You need to login to read this chapter"` error can be thrown.

**Comment by alessandrojean** [comment]
Also, can you please show me how a chapter `a` element looks when you are logged in? (by View Source-Code, because it will show before any javascript code gets executed).

**Comment by TheUnlocked** [comment]
I ran the HTML through a diff checker with JS disabled both when logged in and when not logged in. They're identical.

**Comment by alessandrojean** [comment]
That's bad 😢.

I will see if I can find the way that they do to check if the user is logged in, but I guess that maybe the code checks the cookies. 

If it is the case, then we can filter the chapter list.

**Comment by alessandrojean** [comment]
Does the `div#o_account-links-content` have a attribute `logged_in="true"` when you are logged in, with JS disabled?

**Comment by TheUnlocked** [comment]
It doesn't even exist with JS disabled. :/

**Comment by alessandrojean** [comment]
I figured out. The site does a GET request to `https://www.viz.com/account/refresh_login_links` and uses its HTML contents to populate the links, and it does provide the `logged_in` in the element. 

Maybe we can use this to determine if the user is logged in, see if this works:

```kotlin
override fun chapterListParse(response: Response): List<SChapter> {
    val allChapters = super.chapterListParse(response)

    val newHeaders = headersBuilder()
        .add("X-Requested-With", "XMLHttpRequest")
        .set("Referer", response.request().url().toString())
        .build()
    val loginCheckRequest = GET(REFRESH_LOGIN_LINKS_URL, newHeaders)
    val document = client.newCall(loginCheckRequest).execute().asJsoup()
    val isLoggedIn = document.select("div#o_account-links-content").first()!!.attr("logged_in").toBoolean()

    if (isLoggedIn) {
        return allChapters
            .map { oldChapter ->
                 oldChapter.newCopy().apply { url = url.substringAfterFirst("'").substringBeforeLast("'") }
            }
    }

    return allChapters
        .filter { !it.url.startsWith("javascript") }
}

private fun SChapter.newCopy(): SChapter = SChapter.create().apply {
  name = this.name
  date_upload = this.date_upload
  scanlator = this.scanlator
  url = this.url
}

override fun chapterFromElement(element: Element): SChapter = SChapter.create().apply {
    // ...
    url = element.attr("data-target-url")
}

companion object {
    // ...
    private const val REFRESH_LOGIN_LINKS_URL = "https://www.viz.com/account/refresh_login_links"
}
```

**Comment by TheUnlocked** [comment]
> I figured out. The site does a GET request to https://www.viz.com/account/refresh_login_links and uses its HTML contents to populate the links, and it does provide the logged_in in the element.

Ahh nice. I was trying to do something where I attempt to load a page of the chapter and see if I get an error, but it wasn't terribly successful.

> _Edit_: I am not entirely sure if the `SChapter.newCopy()` will work or if it has any errors.

It wasn't working, though I'm not familiar enough with Kotlin to know why exactly. Something to do with "lateinit" properties?

Regardless, after removing that part everything works, and I'm not that concerned about mutating elements of a local variable list that we don't pass anywhere else.

Thank you so much for working on this!

**Comment by alessandrojean** [comment]
> It wasn't working, though I'm not familiar enough with Kotlin to know why exactly. Something to do with "lateinit" properties?

I don't know, it's supposed to clone the object, allowing that one property is modified on the `map`, but probably there's a better way.

> Regardless, after removing that part everything works

But does it now shows only the free chapters if you are not logged in, and all the chapters if you are?

Can you do the commit with the last changes then, so I can see all the final code and later made some final adjustments?

**Comment by TheUnlocked** [comment]
> But does it now shows only the free chapters if you are not logged in, and all the chapters if you are?

Yes!

**Comment by alessandrojean** [comment]
I appreciate your help! 

I will merge the PR now :)

**Pull request merged**